### ISSUE TITLE
log observers that throw exceptions should not be removed

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#10709]
|<img alt="bruce's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| bruce reported|
|-|-|
|Trac ID|trac#1069|
|Type|defect|
|Created|2005-06-01 22:03:07Z|
|Branch|https://github.com/twisted/twisted/tree/keep-observers-1069|

<details><summary>Searchable metadata</summary>

```
trac-id__1069 1069
type__defect defect
reporter__bruce bruce
priority__highest highest
milestone__ 
branch__branches_keep_observers_1069 branches/keep-observers-1069
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1117663387000000 1117663387000000
changetime__1214398486000000 1214398486000000
version__ 
owner__ 
cc__glyph cc__radix cc__exarkun cc__itamarst cc__jknight cc__bruce cc__titty
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="bruce's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|bruce commented|
|-|-|

```
#!html
<pre>
This is related to issue #989, but is the problem that I've been having.  I
hacked up log.msg and now I can finally see the error that is happening:

2005/06/01 09:54 MDT [Broker,1,10.0.0.155] Traceback (most recent call last):
      File "/home/bruce/nobackup/t/Twisted/twisted/internet/defer.py", line 324,
in _startRunCallbacks
        self._runCallbacks()
      File "/home/bruce/nobackup/t/Twisted/twisted/internet/defer.py", line 337,
in _runCallbacks
        self.result = callback(self.result, *args, **kw)
      File "/home/bruce/nobackup/cvs/lotr/server/sigil/service.py", line 125, in
_ebLogError
        log.err(failure)
      File "/home/bruce/nobackup/t/Twisted/twisted/python/log.py", line 147, in err
        msg(failure=_stuff, isError=1, **kw)
    --- &lt;exception caught here> ---
      File "/home/bruce/nobackup/t/Twisted/twisted/python/log.py", line 221, in msg
        self.observers[i](actualEventDict)
      File "/home/bruce/nobackup/t/Twisted/twisted/python/log.py", line 258, in emit
        text = eventDict['failure'].getTraceback()
      File "/home/bruce/nobackup/t/Twisted/twisted/python/failure.py", line 309,
in getTraceback
        self.printTraceback(file=io, elideFrameworkCode=elideFrameworkCode,
detail=detail)
    exceptions.TypeError: printTraceback() got an unexpected keyword argument
'elideFrameworkCode'

But that doesn't make sense to me .. printTraceback *does* take
elideFrameworkCode as a keyword argument.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Problem is broken signature on twisted.spread.pb.CopyableFailure.printTraceback.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="bruce's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|bruce commented|
|-|-|

```
#!html
<pre>
Doing the obvious fix to that method does the trick.

I can't get PB tracebacks to show up on the calling side, but I'll deal with
that some other time.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Thanks. tracebacks are off by default for security (and efficiency) reasons;
there's a flag somewhere you can set to reenable them.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="bruce's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|bruce commented|
|-|-|

```
#!html
<pre>
The flag is in that class, but even when I set it to True, I don't get them ..
but I'll deal with that in another bug if I can confirm that isn't my own fault
that it happens.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
The PB part of this is fixed in r13923

The logging part still needs work.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
So here are some suggestions on how to solve this problem in general. I think
our strategy is completely wrong, as most exceptions are due to per-message
problems, not real stateful underlying problems; stateful underlying problems
tend to be rare, and probably transient given decently coded initializationcode
(e.g. disk full for filelogobserver).

1. Have a log.DataError raised on bad data which would *not* remove observer.
Log observer formatting code would catch all exceptions and re-raise DataError
on them.
2. Never remove log observers; raised errors are passed only to other observers
so we don't get infinite loops.

Some combination of the two may work as well.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
We could also attempt logging the exception thrown by the observer on that
observer (useful if, as is common, there is only one); if that fails we can just
drop it on the floor or remove the observer depending on what the general policy is.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|titty commented|
|-|-|

```
#!html
<pre>
just been hit by this one.
I can assure you that just ignoring that exception from an observer is *much
better* than removing that observer forever (especially when one uses
log.startLogging(sys.stdout)).
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
I'm gonna say that we should definitely fix this for 2.2.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
decide on something to do so I can implement it :-(
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
1. Separate the keyword handling &amp; string munging into separate function(s)
(this makes it easier for other people to make observers, override behavior, and
hopefully do backwards compat layer when we throw this crap out).

2. Wrap calls to said function in try:/except:, on exception log some sort "bad
input to log.msg" message. These exceptions should not be raised due to bad
interaction with #3 below.

3. If log observer call raises an exception, don't remove observer, just pass
this exception on to other observers. Due to #1 and #2 above, this should only
happen on e.g. disk full errors.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
Ok, re-reading this issue from the beginning, I have no idea what the actual
problem is. First, bruce's initial report was a bug in CopyableFailure, not the
logging system (at least, nobody's described how there's a bug in the logging
system yet). In itamar's last post, he talks about catching exceptions about
message formatting, but those are ALREADY caught, and handled fairly reasonably
(I think; an example of how they're not, if that's the case, would be great). 

Itamar, I'm also not sure which keyword/string munging code you're talking about
separating. FileLogObserver.emit? log.msg?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
Umm, ok, I was mistaking the user-specified formatting code for the default
formatting code. That's braindead.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
Ok, I've fixed up FileLogObserver so that it's impenetrable as far as bad input
goes AFAICT. See branches/log-error-1069.

This included itamar's suggestion #1 and #2 (I think), but doesn't change the
way msg() deals with removing observers. I'm not sure if I'm going to do that
last one, IMO fixing FileLogObserver itself is good enough for a release.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
Ok, and a problem is that since I changed things to use safe_str instead of
str(), UnicodeErrors which were being tested for previously are no longer being
raised, but instead causing the EncodeError to go to the log.

I think that behavior's fine, so the branch now has the test check for the
existence of the UnicodeEncodeError in the log output instead of raised from msg().
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="titty's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|titty commented|
|-|-|

```
#!html
<pre>
Why remove observers on error?

When this bug hit me, it was because of a 'RuntimeError: maximum recursion depth
exceeded', which I then tried to log. This ofcourse also failed (with the same
error..) and it seemed like the program suddenly stopped execution.
Normal prints or whatever didn't work and it took me some time to realize that
it was twisted eating my log output.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Agreed -- we shouldn't remove observers upon error. I've never known it to do 
anything but cause trouble. 

Changes to make the default logger less prone to failure are a good change in any 
case, of course.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
Ok, that's cool, now review the damned branch so I can work on the next one.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun commented|
|-|-|

```
#!html
<pre>
The test methods should be named descriptively.

Overall the branch seems larger than it needs to be, but I don't see anything
wrong with it.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
Ok, thanks exarkun. I've fixed the test names, and described to you off-list why
the branch is so big (% is evil), and merged. I'm moving this bug to "BUG"
priority, meaning I'm not disagreeing with people who say that observers
shouldn't be removed (and I never did, sheesh), I just don't think it's that
important for the 2.2 release especially given this strengthening of
FileLogObserver I just merged. I still may work on the observer semantics
changes soon, but don't bet on it.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_22"></a>|@itamarst commented|
|-|-|

OK, so, since 2.2 we've had at least one bug fixed where observers stopped, and another I just filed. I've had it.

It's time log observers were not removed on exceptions; this should be  highest priority.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

(In [23991]) Branching to 'keep-observers-1069'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun commented|
|-|-|

(In [23992]) do not remove log observers which raise an exception

refs [#10709](https://github.com/twisted/twisted/issues/10709)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun removed owner|
|-|-|

Okay, they're no longer removed.  They won't see their own errors, though.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_27"></a>|@therve commented|
|-|-|

Please merge!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun set status to `closed`|
|-|-|

(In [24018]) Merge keep-observers-1069

Author: exarkun
Reviewer: therve
Fixes: [#10709](https://github.com/twisted/twisted/issues/10709)

Change `LogPublisher` so that it only logs exceptions raised by observers
but does not remove the offending observer.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
